src/coreclr/tools/Common/TypeSystem/Interop/IL/MarshalHelpers.cs

```diff
@@ -228,29 +225,35 @@ public static partial class MarshalHelpers
         {
             elementMarshallerKind = MarshallerKind.Invalid;
 
-            bool isByRef = false;
+            TypeSystemContext context = type.Context;
+            NativeTypeKind nativeType = NativeTypeKind.Default;
+            bool isField = marshallerType == MarshallerType.Field;
+
+            if (marshalAs != null)
+                nativeType = marshalAs.Type;
+
             if (type.IsByRef)
             {
-                isByRef = true;
-
                 type = type.GetParameterType();
 
-                if (!type.IsPrimitive && type.IsValueType && marshallerType != MarshallerType.Field
+                if (type.IsValueType && !type.IsPrimitive && !type.IsEnum && !isField
```

Comment:
Trying to make these conditions consistent and more in sync with CoreCLR logic. This path should be only ever used by managed C++ so it does not matter a while lot anyway.